Applications/Spire/Include/Spire/OrderImbalanceIndicator/FilteredOrderImbalanceIndicatorModel.hpp

```diff
@@ -0,0 +1,121 @@
+#ifndef SPIRE_FILTERED_ORDER_IMBALANCE_INDICATOR_MODEL_HPP
+#define SPIRE_FILTERED_ORDER_IMBALANCE_INDICATOR_MODEL_HPP
+#include "Spire/OrderImbalanceIndicator/OrderImbalanceIndicatorModel.hpp"
+
+namespace Spire {
+
+  //! Represents an OrderImbalanceModel that applies custom filters to the
+  //! data it provides.
+  class FilteredOrderImbalanceIndicatorModel :
+      public OrderImbalanceIndicatorModel {
+    public:
+
+      //! Constructs a FilteredOrderImbalanceIndicatorModel with a given
+      //! source model and specific filters.
+      /*
+        \param source_model The model supplying data to the filtered model.
+        \param filters The functions applied to order imbalances as filters.
+                       Filter functions return true if an imbalance should not
+                       be filtered out.
+      */
+      FilteredOrderImbalanceIndicatorModel(
+        std::shared_ptr<OrderImbalanceIndicatorModel> source_model,
+        std::vector<std::function<bool (const Nexus::OrderImbalance&)>>
+          filters);
+
+      std::tuple<boost::signals2::connection,
+        QtPromise<std::vector<Nexus::OrderImbalance>>>
+        subscribe(const boost::posix_time::ptime& start,
+          const boost::posix_time::ptime& end,
+          const OrderImbalanceSignal::slot_type& slot) override;
+
+    private:
+
+      struct FilteredOrderImbalanceSignalConnection :
+          OrderImbalanceSignalConnection {
+        boost::signals2::scoped_connection m_source_signal_connection;
```

Comment:
No inheritance.

Reply:
Should I just copy the members?

Reply:
Yeah, I'd say remove the ```OrderImbalanceSignalConnection``` entirely.

Reply:
Do you mean to move it to the subclassed models? Or to stop using it entirely?

Reply:
Remove it entirely.

Reply:
The only way I can think of then is to store lambdas that have copied the start and end times, with each lambda representing a subscriber connection. When a new imbalance is published, every lambda gets called and signals only if the new imbalance falls within the start and end time.

Reply:
Just get rid of the ```OrderImbalanceSignalConnection```, your filtered model can still have a struct that represents a connection.

Reply:
But the Cache and Local model use it as well. Is the idea that even if it's duplicated it's an private implementation detail of the inherited models themselves?

Reply:
Yep!